common/error.h

```diff
@@ -23,19 +23,48 @@ struct Success {};
 class [[nodiscard]] Error {
  public:
   // Represents an error state.
-  explicit Error(llvm::Twine message) : message_(message.str()) {
+  explicit Error(llvm::Twine prefix, llvm::Twine location, llvm::Twine message)
```

Comment:
This feels like it's becoming really explorer-specific: does it make sense to move it out of /common? Note I think we may want the simpler `Error` class without this addition for other APIs, but judging by your changes we aren't using it elsewhere right now.

A counter-argument might be that the extra twines are pretty cheap, so using it elsewhere isn't really a problem.

Reply:
We aren't using `Error` with a prefix and location anywhere else, but we are using `Error` and `ErrorOr` with just a string error in code that's shared between the toolchain and explorer.

I do think we want to split this up and move the diagnostic-specific stuff out of `Error` / `ErrorOr`, but I think we also want a more significant redesign of how explorer handles diagnostics that doesn't limit it to producing one error total, and I think it makes sense to make those improvements together. If you'd prefer that we don't do this part now, I can split out the bugfixes for the broken error messages produced by parsing:
```
COMPILATION ERROR: file.carbon:4: COMPILATION ERROR: file.carbon: 4: error message
```
and drop the rest of this patch for now. What do you think?

Reply:
No, I wouldn't parse the string.

We can go with this.